service/matching/reachability.go

```diff
@@ -298,40 +319,54 @@ type reachabilityCache struct {
 	openWFCache    cache.Cache
 	closedWFCache  cache.Cache // these are separate due to allow for different TTL
 	metricsHandler metrics.Handler
+	visibilityMgr  manager.VisibilityManager
 }
 
 func newReachabilityCache(
 	handler metrics.Handler,
+	visibilityMgr manager.VisibilityManager,
 	reachabilityCacheOpenWFExecutionTTL,
-	reachabilityCacheClosedWFExecutionTTL time.Duration) reachabilityCache {
+	reachabilityCacheClosedWFExecutionTTL time.Duration,
+) reachabilityCache {
 	return reachabilityCache{
 		openWFCache:    cache.New(reachabilityCacheMaxSize, &cache.Options{TTL: reachabilityCacheOpenWFExecutionTTL}, handler),
 		closedWFCache:  cache.New(reachabilityCacheMaxSize, &cache.Options{TTL: reachabilityCacheClosedWFExecutionTTL}, handler),
 		metricsHandler: handler,
+		visibilityMgr:  visibilityMgr,
 	}
 }
 
 // Get retrieves the Workflow Count existence value based on the query-string key.
-// It returns !ok if the requested element is not in the cache, or if the cached value is not a boolean.
-func (c *reachabilityCache) Get(queryKey string, open bool) (exists bool, ok bool) {
+func (c *reachabilityCache) Get(ctx context.Context, countRequest manager.CountWorkflowExecutionsRequest, open bool) (exists, hit bool, err error) {
+	// try cache
 	var result interface{}
 	if open {
-		result = c.openWFCache.Get(queryKey)
+		result = c.openWFCache.Get(countRequest)
 	} else {
-		result = c.closedWFCache.Get(queryKey)
+		result = c.closedWFCache.Get(countRequest)
 	}
-	if result == nil {
-		return false, false
+	if result != nil {
+		exists, hit = result.(bool)
+		if hit {
```

Comment:
is it safe to treat second return val as `hit`? doesn't false mean that the return value is not a bool? in that case we should error instead of returning false, false, nil.

Reply:
yes, `hit == false` means the return value is not a bool. As is, the code treats this case as a "miss", queries visibility, and then Puts a correct boolean value into the cache. 

Right now the only way that the cache.Get returns an error is if the visibility count query returns an error. This means that `err != nil` --> `cacheMiss` --> `visibility count query was attempted`. If I change it so that a non-bool hit returns an error, then that would get rid of the `err != nil` --> `visibility count query was attempted` guarantee and confuse the "exit point" categories.

Given that a non-bool entry is pretty much an impossible scenario since `cache.Put` only puts booleans into the cache, I think treating it as a miss makes sense

Reply:
Yeah I think I confused myself. Nevertheless just naming it `ok` instead of `hit` sounds more right to me. It's fine if we treat `ok==false` as miss and not return error.

Reply:
I agree. Done :)